### PR TITLE
feat: customize error formatting and styling

### PR DESCRIPTION
**Pull request by elliot-nelson**
### SUMMARY

* Customize _unexpected error formatting_ with `.unexpectedErrorFormatter()`.
* Customize _unexpected error styling_ with `unexpectedError` style hook.

### DETAILS

```js
require('sywac')
  .command('start', /* ... */)
  .command('stop', /* ... */)
  .unexpectedErrorFormatter(error => error.message)
  .style({ unexpectedError: str => chalk.red(str) })
  .parseAndExit()
```

This PR gives two ways to customize the formatting and/or the styling of the error output of an _unexpected error_ (an error thrown by a `run()` or `check()` handler, or added manually using `context.unexpectedError`).

### CODE REVIEW

File: api.js [hunk 1]
```diff
@@ -40,6 +40,7 @@ class Api {
     this._strictMode = 'strictMode' in opts ? opts.strictMode : false
     this._magicCommandAdded = false
     this._modulesSeen = opts.modulesSeen || []
+    this._errorFormatter = undefined
```
Comment (nexdrew) [inline api.js]: Maybe we could move this to the `configure` method, so there's an alternate way of setting/overriding it without calling `.unexpectedErrorFormatter(fn)`.

Perhaps something like this:
```js
this._errorFormatter = opts.errorFormatter || this._errorFormatter // expects a function
```

File: api.js [hunk 2]
```diff
@@ -258,6 +259,11 @@ class Api {
     return this
   }
 
+  unexpectedErrorFormatter (fn) {
+    this._errorFormatter = fn
```
Comment (nexdrew) [inline api.js]: I kinda wish the name of the variable more closely resembled the name of the Api method, but I also prefer shorter names to longer ones.

I don't know, this is probably fine. (No action needed.)

File: api.js [hunk 3]
```diff
@@ -710,7 +716,14 @@ class Api {
...
       fsLib: this.fsLib,
-      state
+      errorFormatter: this._errorFormatter,
+      state,
+
+      // Pass all help options (styling) down to the Context constructor. For the most
+      // part, these options will be ignored, as the Context usually does not care about
+      // styling - but it will pick up individual options if appropriate (such as
+      // `styleUnexpectedError` if there is unexpected error).
+      ...this.helpOpts
```
Comment (nexdrew) [inline api.js]: It seems a little odd/unnecessary/inefficient to me to pass _all_ `helpOpts` to the Context, but I guess it doesn't hurt anything. (No action needed.)

File: context.js [hunk 4]
```diff
@@ -144,7 +146,12 @@ class Context {
 
   unexpectedError (err) {
     this.errors.push(err)
-    this.output = String((err && err.stack) || err)
+    const message = typeof this.errorFormatter === 'function'
+      ? String(this.errorFormatter(err))
+      : String((err && err.stack) || err)
+    this.output = typeof this.styleUnexpectedError === 'function'
+      ? this.styleUnexpectedError(message)
+      : message
```
Comment (nexdrew) [inline context.js]: I wish there was a way to do this where we didn't have to split style hook executions between different locations, but I can't think of a better way of doing it than this.

The only _potential_ problem I see is that this doesn't leverage the `all` style hook (as a final wrapper for all other styled strings), but it wasn't doing that before, and I'm not sure if a CLI author would expect the `all` hook to apply to a formatted error message or not (currently it only applies to help text and expected CLI messages). What do you think?

File: test/test-api.js [hunk 5]
```diff
@@ -222,6 +222,38 @@ tap.test('api > attempt to get unmapped type returns null (don\'t do this)', t =
...
+  t.equal(result1.output, 'An Error')
+
+  // Customized styling
+  api.style({ unexpectedError: str => `[[${str}]]` })
+  const result2 = await api.parse()
+  t.equal(result2.output, '[[An Error]]')
+
+  t.end()
+})
+
```
Comment (nexdrew) [inline test/test-api.js:256]: Might be good to also have a test that applies formatting/styling to an error thrown from a (nested) command.

File: api.js (hunk 1 above)
Comment (nexdrew) [inline api.js]: On another note, when I first saw this, I thought the `_errorFormatter` property should be propagated down to child Api instances via `newChild`, but then I realized that it's only applied at the top-level since that's where the single Context instance (per parse) is initialized.

When documenting this, we should note that both `.unexpectedErrorFormatter(fn)` _and_ the `unexpectedError` style hook **can only be applied at the top-level** (i.e. cannot be applied/overridden by command configuration), because a single "top-level" Context instance is used for all (potentially nested) command levels.

**Review by nexdrew — COMMENTED** [review]
Awesome, thanks for doing this! I appreciate that you were able to keep the implementation simple and straightforward.

I just have a handful of comments/questions. Let me know what you think. Thanks!

File: api.js (hunk 3 above)
Comment (elliot-nelson) [reply api.js]: I feel like, once you take the plunge to say "we sometimes require styling in `Api`", you should just pass it all.

I'd be totally comfortable including it nested if you felt that was better encapsulated -- i.e. instead of `...helpOpts` pass in `helpOpts` as a key.

File: context.js (hunk 4 above)
Comment (elliot-nelson) [reply context.js]: I guess it depends what you're doing with `all`.  Possible use case:

```js
all: text => {
    // Example weird use case: always return output as a JSON blob containing lines as an array,
    // for consumption by some internal API.
    return JSON.stringify({ lines: text.split('\n') })
}
```

If this is how I was using sywac, then I _would_ want any output error messages to be run through it as well (although, in this case, I'm probably using `parse()` and could just examine `.errors` instead of the output).

My concern about just wrapping it with `all` is that you would potentially run afoul of `all` handers written just to handle the expected help text screen and not a generic error.  I almost feel you need to pass a second boolean option to tell the function whether this is a "help text" context or "unexpected error" context... but if you did that, and the user actually uses the boolean to make a decision, then I feel you could just as easily require them to put the logic in the unexpected error handler:

```js
all: text => {
    return JSON.stringify({ lines: text.split('\n') })
},
unexpectedError: text => {
    return JSON.stringify({ lines: text.split('\n') })
},
```

TLDR: This is a good point but I think I land on the side of "all does not apply to unexpected errors, because that's more flexible than the alternative".

File: api.js (hunk 3 above)
Comment (nexdrew) [reply api.js]: > I'd be totally comfortable including it nested if you felt that was better encapsulated -- i.e. instead of `...helpOpts` pass in `helpOpts` as a key.

My gut instinct tells me this would be better than how it currently is, though I can't objectively defend that.

If you're willing to make that change, I'd appreciate it.

File: context.js (hunk 4 above)
Comment (nexdrew) [reply context.js]: I think I agree with you. Thanks for the feedback!

File: api.js (hunk 1 above)
Comment (nexdrew) [reply api.js]: In preparation for implementing #65, perhaps we could go ahead and throw an error from `.unexpectedErrorFormatter(fn)` if it's called on a child Api instance?

Just trying to think in terms of setting a precedent for the future. Otherwise, I'm completely fine if the calls simply get ignored in the initial implementation of this feature.

**Comment by elliot-nelson** [comment]
New commit pushed:

 - Changed all references to the new formatter to `unexpectedErrorFormatter` (although it's longer, I think it's less confusing to refer to a value by the same name everywhere we can).
 - Pass `helpOpts` as a property instead of collapsing into the rest of the options for Context.
 - Added a test where an error thrown by a subcommand is styled & formatted.

File: api.js (hunk 1 above)
Comment (elliot-nelson) [reply api.js]: My preference is to start with no error for now -- in my opinion, if people complained about being bitten by this bug, then the "right fix" isn't to prevent people from passing the error formatter to subcommands, it's to fix it so it works as the user was assuming.

(That's out of scope for this PR, but I think it's possible if we cared enough -- we'd need to capture an unexpected error thrown at the `Command` level and capture the error formatter from that context, and then ensure that one got used if and when an unexpected error was printed.  I consider this a "potential future todo" that's better than just preventing people from using the option.)
